src/main/drivers/bus_spi.c

```diff
@@ -259,54 +397,204 @@ uint16_t spiCalculateDivider(uint32_t freq)
+        case BUS_BUSY:
+            // Repeat the last DMA segment
+            bus->curSegment--;
+            // Reinitialise the cached init values as segment is not progressing
+            spiPrivInitStream(dev, true);
+            break;
+
+        case BUS_ABORT:
+            bus->curSegment = (busSegment_t *)NULL;
+            return;
+
+        case BUS_READY:
+        default:
+            // Advance to the next DMA segment
+            break;
+        }
+    }
+
+    // Advance through the segment list
+    bus->curSegment++;
+
+    if (bus->curSegment->len == 0) {
+        // The end of the segment list has been reached, so mark transactions as complete
+        bus->curSegment = (busSegment_t *)NULL;
+    } else {
+        // After the completion of the first segment setup the init structure for the subsequent segment
+        if (bus->initSegment) {
+            spiPrivInitStream(dev, false);
+            bus->initSegment = false;
+        }
+
+        // Launch the next transfer
+        spiPrivStartDMA(dev);
+
+        // Prepare the init structures ready for the next segment to reduce inter-segment time
+        spiPrivInitStream(dev, true);
+    }
 }
 
-bool spiBusTransactionTransfer(const busDevice_t *bus, const uint8_t *txData, uint8_t *rxData, int length)
+
+bool spiSetBusInstance(extDevice_t *dev, uint32_t device, resourceOwner_e owner)
 {
-    spiBusTransactionSetup(bus);
-    return spiBusTransfer(bus, txData, rxData, length);
+    if (device > SPIDEV_COUNT) {
+        return false;
+    }
+
+    dev->bus = &spiBusDevice[SPI_CFG_TO_DEV(device)];
+
+    if (dev->bus->busType == BUSTYPE_SPI) {
+        // This bus has already been initialised
+        dev->bus->deviceCount++;
+        return true;
+    }
+
+    busDevice_t *bus = dev->bus;
+
+    bus->busType_u.spi.instance = spiInstanceByDevice(SPI_CFG_TO_DEV(device));
+
+    if (bus->busType_u.spi.instance == NULL) {
+        return false;
+    }
+
+    bus->busType = BUSTYPE_SPI;
+    bus->useDMA = false;
+    bus->useAtomicWait = false;
+    bus->deviceCount = 1;
+    bus->owner = owner;
+    bus->initTx = &dev->initTx;
+    bus->initRx = &dev->initRx;
+
+    return true;
 }
 
-bool spiBusTransactionWriteRegister(const busDevice_t *bus, uint8_t reg, uint8_t data)
+void spiInitBusDMA()
 {
-    spiBusTransactionSetup(bus);
-    return spiBusWriteRegister(bus, reg, data);
+    uint32_t device;
+#ifdef STM32F4
+    /* Check https://www.st.com/resource/en/errata_sheet/dm00037591-stm32f405407xx-and-stm32f415417xx-device-limitations-stmicroelectronics.pdf
+     * section 2.1.10 which reports an errata that corruption may occurs on DMA2 if AHB peripherals (eg GPIO ports) are
+     * access concurrently with APB peripherals (eg SPI busses). Bitbang DSHOT uses DMA2 to write to GPIO ports. If this
+     * is enabled, then don't enable DMA on an SPI bus using DMA2
+     */
+    const bool dshotBitbangActive = isDshotBitbangActive(&motorConfig()->dev);
```

Comment:
OK, in the same way that `spiSetAtomicWait()` is used to record that a device is using the bus in a particular way, it might make sense to have a `spiSetAHBDMA()`, but then the motor driver would have to make calls to the SPI bus driver which it would not otherwise use. The `spiInitBusDMA()` routine is specifically concerned about SPI DMA of course, but it is actually checking that it's OK to use DMA on a particular SPI bus based on stream availability, and this errata. I can see merits in doing this either way. The motor driver is arguably at the same level as the bus driver as they are both directly programming DMA streams. Were the motor driver an SPI device I'd agree, but it's not.

Further thoughts?

Reply:
@SteveCEvans invert the dependency.  have the motor driver tell the bus system that it needs something special.   "Tell don't ask" applies here.